### ISSUE TITLE
Maximum call stack size exceeded

### ISSUE THREAD

**Issue by mysticatea** [bcoe/nyc#74]
Hello.

I'm trying to use nyc with Babel.
But I got a stackoverflow error.
What am I doing wrong?

```
> nyc --require babel-register node_modules/.bin/mocha test.js

path.js:174
  resolvedTail = normalizeArray(resolvedTail.split(/[\\\/]+/),
                                             ^

RangeError: Maximum call stack size exceeded
    at String.split (native)
    at Object.win32.resolve (path.js:174:46)
    at Object.win32.relative (path.js:259:16)
    at getRelativePath (C:\Users\t-nagashima.AD\Documents\GitHub\sand\node_modules\babel-register\lib\node.js:72:28)
    at shouldIgnore (C:\Users\t-nagashima.AD\Documents\GitHub\sand\node_modules\babel-register\lib\node.js:123:12)
    at require.extensions.(anonymous function) (C:\Users\t-nagashima.AD\Documents\GitHub\sand\node_modules\babel-register\lib\node.js:137:9)
    at requireHook (C:\Users\t-nagashima.AD\Documents\GitHub\sand\node_modules\nyc\index.js:139:7)
    at require.extensions.(anonymous function) (C:\Users\t-nagashima.AD\Documents\GitHub\sand\node_modules\babel-register\lib\node.js:138:7)
    at requireHook (C:\Users\t-nagashima.AD\Documents\GitHub\sand\node_modules\nyc\index.js:139:7)
    at require.extensions.(anonymous function) (C:\Users\t-nagashima.AD\Documents\GitHub\sand\node_modules\babel-register\lib\node.js:138:7)
```

----

Windows 7 64bit
Node v4.2.2
npm v3.5.0

package.json

```json
{
  "name": "nyc-example",
  "version": "0.0.0",
  "private": true,
  "description": "Please run `npm install` and `npm test`.",
  "scripts": {
    "test": "nyc --require babel-register node_modules/.bin/mocha test.js"
  },
  "devDependencies": {
    "babel-core": "^6.3.2",
    "babel-preset-es2015": "^6.1.18",
    "babel-register": "^6.3.2",
    "mocha": "^2.3.4",
    "nyc": "^4.0.1"
  }
}
```

.babelrc

```json
{
  "presets": ["es2015"]
}
```

test.js

```js
class Foo {
}

console.log(Foo);
```

**Comment by novemberborn** [comment]
@mysticatea does it crash if you run `mocha --require babel-register test.js`?

**Comment by mysticatea** [comment]
It does not crash.

```
> mocha --require babel-register test.js

[Function: Foo]


  0 passing (1ms)
```

**Comment by novemberborn** [comment]
Works for me with Node 4.2.1 and 5.0 on OS X, which makes me think this is a Windows issue. Is there anything interesting in the longer stack trace? There's repetition here:

```
    at requireHook (C:\Users\t-nagashima.AD\Documents\GitHub\sand\node_modules\nyc\index.js:139:7)
    at require.extensions.(anonymous function) (C:\Users\t-nagashima.AD\Documents\GitHub\sand\node_modules\babel-register\lib\node.js:138:7)
    at requireHook (C:\Users\t-nagashima.AD\Documents\GitHub\sand\node_modules\nyc\index.js:139:7)
    at require.extensions.(anonymous function) (C:\Users\t-nagashima.AD\Documents\GitHub\sand\node_modules\babel-register\lib\node.js:138:7)
```

but then it crashes on `String.split`?!

```
    at String.split (native)
    at Object.win32.resolve (path.js:174:46)
    at Object.win32.relative (path.js:259:16)
    at getRelativePath (C:\Users\t-nagashima.AD\Documents\GitHub\sand\node_modules\babel-register\lib\node.js:72:28)
    at shouldIgnore (C:\Users\t-nagashima.AD\Documents\GitHub\sand\node_modules\babel-register\lib\node.js:123:12)
    at require.extensions.(anonymous function) (C:\Users\t-nagashima.AD\Documents\GitHub\sand\node_modules\babel-register\lib\node.js:137:9)
```

@bcoe do you have access to a Windows box to confirm this on?

**Comment by mysticatea** [comment]
```
    at requireHook (C:\Users\t-nagashima.AD\Documents\GitHub\sand\node_modules\nyc\index.js:139:7)
    at require.extensions.(anonymous function) (C:\Users\t-nagashima.AD\Documents\GitHub\sand\node_modules\babel-register\lib\node.js:138:7)
    at requireHook (C:\Users\t-nagashima.AD\Documents\GitHub\sand\node_modules\nyc\index.js:139:7)
    at require.extensions.(anonymous function) (C:\Users\t-nagashima.AD\Documents\GitHub\sand\node_modules\babel-register\lib\node.js:138:7)
```

This is a stackoverflow, so there is possibility of clashing in every function call.

- https://github.com/bcoe/nyc/blob/v4.0.1/index.js#L139
- https://github.com/babel/babel/blob/v6.3.2/packages/babel-register/src/node.js#L108

Hmm, Babel seems to call the old extension here, but the old extension is nyc and nyc is calling Babel. So looks there is a potential infinity loop.

`filename` is `C:\Users\t-nagashima.AD\Documents\GitHub\sand\node_modules\mocha\bin\mocha` so it's ignored.

**Comment by bcoe** [comment]
@mysticatea could you give the current `master` branch a shot, there have been some changes to the require-hook that I think might fix the infinite you're seeing.

Disclaimer, I have yet to get subprocess spawning fully functional in Windows.

**Comment by mysticatea** [comment]
Ok, I will try it later.

**Comment by bcoe** [comment]
@mysticatea I tested the master branch of `nyc` on Windows, and it worked great with a couple changes:

<img width="883" alt="screen shot 2015-12-06 at 5 18 09 pm" src="https://cloud.githubusercontent.com/assets/194609/11616984/5c91a8bc-9c3d-11e5-9385-cf024280aabb.png">

1) `test.js` is ignored by default, so you should change the filename to something like `foo.js`.
2) `nyc` currently has trouble capturing exit events in windows, so rather than running `mocha` I ran the `_mocha` bin (which does not spawn a subprocess).

`nyc --require babel-register node_modules/.bin/_mocha test.js`

3) by default the babel register hook will be called for each module in the `node_modules` folder, you should probably configure babel to ignore this folder.

**Comment by mysticatea** [comment]
Thank you for the verification and the information!
I also verified that `master` branch works.
I'm looking forward to the next release. :)

**Comment by bcoe** [comment]
@mysticatea I've released `nyc@5.0.0` with these fixes for Windows, I've also patched Windows so that it works without needing to reference `_mocha`:

https://github.com/bcoe/nyc/pull/87

Your original `node_modules/.bin/mocha` should work now.

**Comment by mysticatea** [comment]
Thank you so much!
